rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,429 @@
+
+```yaml
+piv:
+- name: YubiKey 5 Nano (5.2.7) [OTP+FIDO+CCID] # PIV card name
+  id: xxxxxxxx                                 # unique ID
+  slot: 9a                                     # PIV slot
+```
+
+Note: for yubikey, the unique ID will be the card's serial number. Different implementations of PIV may have alternative methods for identifying unique cards
+
+##### Check the public key
+
+It's possible that after login, the user overwrites the PIV slot used to login. To catch this, future calls to `tsh` will retrieve the public key from the PIV slot and check if it matches the public key on the user's TLS certificate (`~/.tsh/keys/proxy/user-x509.pem`). If it doesn't match, then the user will see an error and get prompted to relogin.
+
+```
+> tsh ls
+ERROR: PIV private key does not match login certificates.
+// Trigger relogin
+```
+
+##### Concurrent login sessions
+
+`tsh` should support multiple concurrent login sessions for different proxies. This can be done by specifying a different slot for each concurrent login session, allowing up to 24 concurrent PIV login sessions. However, this requires that each user maintain a mental map of which slots are in use to avoid overwriting an existing login session.
+
+To move the burden off of the user, `tsh login --piv` will also create a self-signed certificate with login metadata attached and add it to the slot.
+
+```
+Slot 9a:
+    Certificate:
+        Subject DN:	CN = yubikey, O = tsh, OU = v10.0.0, L = proxy.example.com:user
+```
+
+Now, future calls to `tsh login --piv --piv-slot=9a` can retrieve the certificate from the slot and decide what to do:
+ - if there is no certificate, overwrite the slot
+ - if there is a tsh-generated certificate that matches the current user and proxy, overwrite the slot
+ - if there is a tsh-generated certificate that doesn't match the current user and proxy, prompt the user
+  ```
+  > tsh login --piv
+  PIV slot 9a is currently in use by another tsh login session:
+    proxy: proxy.example.com
+    user: username
+  Would you like to overwrite this slot? (y/N):
+  ```
+ - if there is a non-tsh-generated certificate, overwrite the slot, prompt the user
+  ```
+  > tsh login --piv
+  PIV slot 9a may be in use by another program.
+  Would you like to overwrite this slot? (y/N):
+  ```
+
+We will also provide the user with the `--piv-slot-overwrite` flag to skip the prompts above.
+
+Additionally, users can provide `--piv-slot=next` to use the next available slot, so that they don't need to manually find an open slot.
```

Comment:
Would it make sense to make this default behavior? I think we should try to abstract users from slots, etc. if we can help it, I'm pretty sure an average user would have no idea what slots are and which one they should pick.